### PR TITLE
Fix docstrings, error messages

### PR DESCRIPTION
**Pull request by jbrockmendel**
The discussion in #19558 makes me think getting PeriodArray and DatetimeArray implemented sooner rather than later will make things easier for other goals.  In preparation for that, this PR goes through (most) of the methods that will need to be ported and adapts their error messages to be class-agnostic, and to use `str.format` syntax.

Also fixes a couple of copy/paste typos in docstrings.

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -670,8 +670,9 @@ def __add__(self, other):
             elif isinstance(self, TimedeltaIndex) and isinstance(other, Index):
                 if hasattr(other, '_add_delta'):
                     return other._add_delta(self)
-                raise TypeError("cannot add TimedeltaIndex and {typ}"
-                                .format(typ=type(other)))
+                raise TypeError("cannot add {cls} and {typ}"
+                                .format(cls=type(self).__name__,
+                                        typ=type(other).__name__))
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimelike.py]: I am not sure this is an improvement in robustness, as *if* series would use the same code, this 'cls' would become either "Series" or "TimedeltaArray", and "Series" is not that informative, and "TimedeltaArray" should not bubble up to user when using a Series (that would be an internal implementation detail).

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -804,7 +806,7 @@ def shift(self, n, freq=None):
 
         Returns
         -------
-        shifted : DatetimeIndex
+        shifted : type(self)
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimelike.py]: same here

File: pandas/core/indexes/datetimes.py [hunk 3]
```diff
@@ -106,15 +106,15 @@ def _dt_index_cmp(opname, cls, nat_result=False):
...
+        binop = getattr(Index, opname)
 
         if isinstance(other, (datetime, compat.string_types)):
             if isinstance(other, datetime):
                 # GH#18435 strings get a pass from tzawareness compat
                 self._assert_tzawareness_compat(other)
 
             other = _to_m8(other, tz=self.tz)
-            result = func(other)
+            result = binop(self, other)
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimes.py]: why this change?

**Review by jorisvandenbossche — COMMENTED** [review]
Added a few comments, but in general, I would defer this to when we *actually* changed things instead of now already trying to be "prepared" or "robust" for such changes without knowing how it will exactly be done in practice.

I would rather put effort in actually finding out how we can such a refactor to dispatch ops to Array-likes, backing both ops for Series/Block and Index, but that is something you have to sync closely with @TomAugspurger to make sure you don't conflict with what he is doing (and I think help is certainly welcome! we just need to make sure we are not doing duplicate work)

**Comment by TomAugspurger** [comment]
Haven't looked closely yet, but I haven't touched datetimeindex yet.

Including the correct class name in the error message has been an issue, so those changes should be worthwhile.

Should be able to look closer in a few days.

File: pandas/core/indexes/datetimelike.py [hunk 4]
```diff
@@ -794,7 +796,7 @@ def isin(self, values):
 
     def shift(self, n, freq=None):
         """
-        Specialized shift which produces a DatetimeIndex
+        Specialized shift which produces a new object of the same type as self
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: This is a good note.  But the reason for this change isn't future-looking; this is also the docstring for `TimedeltaIndex.shift`.

File: pandas/core/indexes/datetimes.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: Because `DatetimeIndex.__mro__` has 12 entries, so with the standard `super` usage it isn't obvious where the method is defined.

**Comment by jbrockmendel** [comment]
Just pushed a fix for the `shift` docstring that re-evaluates it within `TimedeltaIndex` and `DatetimeIndex`.  I'm wondering if there is a metaclass (or hopefully a lighter-weight option) way to do this without having to re-define the function in each subclass.  Could that be what `_decorators.docstring_wrapper` (which is not used) is for?

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/indexes/datetimelike.py]: The above is *always* a TimedeltaIndex given the elif statement above

File: pandas/core/indexes/datetimelike.py [hunk 5]
```diff
@@ -705,8 +706,9 @@ def __sub__(self, other):
                 return self._sub_offset_array(other)
             elif isinstance(self, TimedeltaIndex) and isinstance(other, Index):
                 if not isinstance(other, TimedeltaIndex):
-                    raise TypeError("cannot subtract TimedeltaIndex and {typ}"
-                                    .format(typ=type(other).__name__))
+                    raise TypeError("cannot subtract {cls} and {typ}"
+                                    .format(cls=type(self).__name__,
+                                            typ=type(other).__name__))
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimelike.py]: same here

File: pandas/core/indexes/datetimes.py [hunk 6]
```diff
@@ -826,18 +826,19 @@ def _add_datelike(self, other):
...
 
     def _sub_datelike(self, other):
         # subtract a datetime from myself, yielding a TimedeltaIndex
         from pandas import TimedeltaIndex
         if isinstance(other, DatetimeIndex):
             # require tz compat
             if not self._has_same_tz(other):
-                raise TypeError("DatetimeIndex subtraction must have the same "
-                                "timezones or no timezones")
+                raise TypeError("{cls} subtraction must have the same "
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimes.py]: Same here, this will always be subtraction of DatetimeIndexes ?

File: pandas/core/indexes/datetimes.py [hunk 7]
```diff
@@ -854,8 +855,9 @@ def _sub_datelike(self, other):
                 result = self._maybe_mask_results(result,
                                                   fill_value=libts.iNaT)
         else:
-            raise TypeError("cannot subtract DatetimeIndex and {typ}"
-                            .format(typ=type(other).__name__))
+            raise TypeError("cannot subtract {cls} and {typ}"
+                            .format(cls=type(self).__name__,
+                                    typ=type(other).__name__))
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimes.py]: same here?

File: pandas/core/indexes/timedeltas.py [hunk 8]
```diff
@@ -59,23 +59,25 @@ def _td_index_cmp(opname, cls, nat_result=False):
     """
 
     def wrapper(self, other):
-        msg = "cannot compare a TimedeltaIndex with type {0}"
-        func = getattr(super(TimedeltaIndex, self), opname)
+        msg = "cannot compare a {cls} with type {typ}"
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/timedeltas.py]: same here? `self` is always a TimedeltaIndex

File: pandas/core/indexes/datetimelike.py [hunk 9]
```diff
@@ -794,7 +796,7 @@ def isin(self, values):
 
     def shift(self, n, freq=None):
         """
-        Specialized shift which produces a DatetimeIndex
+        Specialized shift which produces a %(klass)s
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimelike.py]: Instead of doing the overriding in the subclass just to fix this name, we can maybe reword the docstring to be clear enough? 
Eg mention it works for Datetime- and TimedeltaIndex?

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: That was the idea in the earlier version.  I'm open to suggestions.

File: pandas/core/indexes/datetimes.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: It is now, but one of the on-deck PRs will have ndarray[datetime64] go down this path too to fix:

```
dti = pd.date_range('2016-01-01', periods=3)
arr = np.array(['2015-01-01', '2015-02-02', '2015-03-03'], dtype='datetime64[ns]')
>>> dti - arr
DatetimeIndex(['1971-01-01', '1970-12-01', '1970-11-03'], dtype='datetime64[ns]', freq=None)
```

Comment (jreback) [reply pandas/core/indexes/datetimes.py]: how about instead of using ``type(self).__name__`` we make a little function which can format these. I agree with @jorisvandenbossche point here, in the case of an Index type the dtype is clear, but not for a Series.

File: pandas/core/indexes/datetimes.py [hunk 10]
```diff
@@ -901,6 +903,11 @@ def _sub_datelike_dti(self, other):
             new_values[mask] = libts.iNaT
         return new_values.view('i8')
 
+    @Substitution(klass='DatetimeIndex')
+    @Appender(DatetimeIndexOpsMixin.shift.__doc__)
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: this should be a shared doc-string instead (and same for the other things you added). this is the existing pattern.

File: pandas/core/indexes/timedeltas.py (hunk 8 above)
Comment (jreback) [reply pandas/core/indexes/timedeltas.py]: can you add some tests which lock down these error messages? IOW the types that are *supposed* to appear

File: pandas/core/indexes/timedeltas.py [hunk 11]
```diff
@@ -409,7 +412,7 @@ def _add_datelike(self, other):
             result = checked_add_with_arr(i8, other.value,
                                           arr_mask=self._isnan)
             result = self._maybe_mask_results(result, fill_value=iNaT)
-        return DatetimeIndex(result, name=self.name, copy=False)
```
Comment (jreback) [inline pandas/core/indexes/timedeltas.py]: what changed here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/timedeltas.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/core/indexes/timedeltas.py]: It just got indented (next line in the diff) b/c this branch is never reached _except_ for in the indented case.

File: pandas/core/indexes/datetimes.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: This seems worthwhile but a pretty low priority.  Closing for now.
